### PR TITLE
Attempts to not mutate FunctionInstance on export

### PR DESCRIPTION
**Pull request by codefromthecrypt**
OK this fails on imported calls on both interpreter and JIT (in the enginetest which is missing something). Also JIT fails on another test due to subtracting from a zero PC

### CODE REVIEW

File: internal/testing/enginetest/enginetest.go [hunk 1]
```diff
@@ -260,112 +266,125 @@ func RunTestModuleEngine_Call_HostFn(t *testing.T, et EngineTester) {
...
 }
 
 const (
-	wasmFnName     = "wasm_div_by"
-	hostFnName     = "host_div_by"
-	callHostFnName = "call->" + hostFnName
+	wasmFnName               = "wasm_div_by"
+	hostFnName               = "host_div_by"
+	callHostFnName           = "call->" + hostFnName
+	callImportCallHostFnName = "call_import->" + callHostFnName
```
Comment (codefromthecrypt) [inline internal/testing/enginetest/enginetest.go:387]: all tests fail on this. maybe there's a setup issue

**Comment by codefromthecrypt** [comment]
@mathetake my brain is mush so if you can take a look and maybe push a commit to get expected behavior on a module-defined function calling an imported one, much obliged. I can't see what's wrong and really want to be able to test moduleEngine independently inclusive of errors, without relying on Store side-effects, to preserve time in the future.

**Comment by codefromthecrypt** [comment]
almost all code is here https://github.com/tetratelabs/wazero/pull/438

**Comment by codefromthecrypt** [comment]
ok I rebased this over #438 to show what's different.

**Comment by mathetake** [comment]
oh what's failing now?

**Comment by mathetake** [comment]
ah CI isn't running ok 😄

**Comment by codefromthecrypt** [comment]
clicked ready for review in case that kicks CI

**Comment by codefromthecrypt** [comment]
oh it is because I put this against the other branch

**Comment by codefromthecrypt** [comment]
https://github.com/tetratelabs/wazero/actions/runs/2088507390

File: internal/testing/enginetest/enginetest.go [hunk 2]
```diff
@@ -402,7 +431,30 @@ func setupCallTests(t *testing.T, e wasm.Engine) (*wasm.ModuleInstance, wasm.Mod
...
+		Type:  ft,
+		Body:  []byte{wasm.OpcodeLocalGet, 0, wasm.OpcodeCall, 0 /* only one imported function */, wasm.OpcodeEnd},
+		Index: 1, // after import
+	}
+	addFunction(importing, callImportCallHostFnName, callImportedHostFn)
+
+	// Compile the importing module
+	importingMe, err := e.NewModuleEngine(importing.Name, importedFunctions, importing.Functions, nil, nil)
+	require.NoError(t, err)
+	linkModuleToEngine(importing, importedMe)
```
Comment (mathetake) [inline internal/testing/enginetest/enginetest.go:452]: This must be `importingMe`, not `importedMe`.
```suggestion
	linkModuleToEngine(importing, importingMe)
```

Comment (mathetake) [reply internal/testing/enginetest/enginetest.go:452]: still failure remains, but this makes the error constsitent between Iterpreter and JIT

**Comment by mathetake** [comment]
oh yeah this solves the outstanding issues except ExpecteError=`TODO:` cases 😄

**Comment by mathetake** [comment]
fix pushed!

**Comment by mathetake** [comment]
ok cool passed on all platforms https://github.com/tetratelabs/wazero/runs/5813115061?check_suite_focus=true

File: internal/testing/enginetest/enginetest.go [hunk 3]
```diff
@@ -449,7 +461,7 @@ func setupCallTests(t *testing.T, e wasm.Engine) (*wasm.ModuleInstance, wasm.Mod
 	// Compile the importing module
 	importingMe, err := e.NewModuleEngine(importing.Name, importedFunctions, importing.Functions, nil, nil)
 	require.NoError(t, err)
-	linkModuleToEngine(importing, importedMe)
```
Comment (codefromthecrypt) [inline internal/testing/enginetest/enginetest.go]: hah definitely appreciate the second eyes!
